core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1088,7 +1088,17 @@ class ReplicaManager(val config: KafkaConfig,
         quota = quota,
         clientMetadata = clientMetadata)
       if (isFromFollower) updateFollowerFetchState(replicaId, result)
-      else result
+      else {
+        // the fetch is from a consumer, sample the age of consumed data
+        if (random.nextDouble() <= config.liConsumerFetchSampleRatio) {
```

Comment:
Just to check:  are there no reports of bias in Scala's PRNG? For example, if it tended not to produce small values, or if it overproduced them, this wouldn't work as intended. (Normally I'd expect everyone would pick something decent since there's literally decades of literature on it, but after seeing the poor results of their split algorithm, I'm not so sure anymore.)

Reply:
The doc says it's uniformly distributed between 0 and 1.0.
https://www.scala-lang.org/api/2.12.4/scala/util/Random.html#nextBoolean():Boolean
So I'll trust the doc unless proven otherwise.